### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by UnfforU**
Pull Request for LibraryWebAPI v1.0 

Create basic services and controllers.

### CODE REVIEW

File: LibraryWebAPI/Controllers/LibraryController.cs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        public LibraryController(ILibraryService libaryService)
+        {
+            _libraryService = libaryService;
+        }
+
+        // GET: api/Library
+        [HttpGet]
+        public async Task<ActionResult<List<Library>>> GetLibraries()
+        {
+            return await _libraryService.GetAllLibraries();
```
Comment (MaltsevaSoftline) [inline LibraryWebAPI/Controllers/LibraryController.cs:28]: If method is async it's a good practice to call it like GetAllLibrariesAsync. Good ref - linq methods

File: LibraryWebAPI/Controllers/UserController.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    {
+        private readonly IUserService _userService;
+        public UserController(IUserService userService)
+        {
+            _userService = userService;
+        }
+
+        //POST api/User
+        [HttpPost]
+        public async Task<List<User>> AddUser(UserDTO userDTO)
```
Comment (MaltsevaSoftline) [inline LibraryWebAPI/Controllers/UserController.cs:28]: No IActionResult

File: LibraryWebAPI/Models/Author.cs [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+namespace LibraryWebAPI.Models;
+
+public partial class Author
+{
+    public Guid AuthorId { get; set; }
+
+    public string Name { get; set; } = null!;
+
+    public virtual ICollection<Book> Books { get; set; } = new List<Book>();
```
Comment (MaltsevaSoftline) [inline LibraryWebAPI/Models/Author.cs:12]: Do you need this property?

File: LibraryWebAPI/Models/Book.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace LibraryWebAPI.Models;
+
+public partial class Book
+{
+    public Guid BookId { get; set; }
+
+    public string Name { get; set; } = null!;
```
Comment (MaltsevaSoftline) [inline LibraryWebAPI/Models/Book.cs:10]: ?

File: LibraryWebAPI/Models/LibraryContext.cs [hunk 5]
```diff
@@ -0,0 +1,130 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace LibraryWebAPI.Models;
+
+public partial class LibraryContext : DbContext
+{
+    public LibraryContext()
```
Comment (MaltsevaSoftline) [inline LibraryWebAPI/Models/LibraryContext.cs:8]: Do you need an empty constructor?

File: LibraryWebAPI/Models/LibraryContext.cs [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+    public virtual DbSet<Author> Authors { get; set; }
+
+    public virtual DbSet<Book> Books { get; set; }
+
+    public virtual DbSet<Library> Libraries { get; set; }
+
+    public virtual DbSet<User> Users { get; set; }
+
+    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
```
Comment (MaltsevaSoftline) [inline LibraryWebAPI/Models/LibraryContext.cs:25]: Do you need this method?

File: LibraryWebAPI/Models/LibraryContext.cs [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    {
+        modelBuilder.Entity<Author>(entity =>
+        {
+            entity.ToTable("author");
+
+            entity.Property(e => e.AuthorId)
+                .HasDefaultValue(new Guid())
+                .HasColumnName("authorId");
+            entity.Property(e => e.Name)
+                .HasMaxLength(50)
```
Comment (MaltsevaSoftline) [inline LibraryWebAPI/Models/LibraryContext.cs:40]: Name might be more than 50 chars

File: LibraryWebAPI/Models/LibraryContext.cs [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+                .IsUnicode(false)
```
Comment (MaltsevaSoftline) [inline LibraryWebAPI/Models/LibraryContext.cs:41]: and your customer might be from China

File: LibraryWebAPI/Models/LibraryContext.cs [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+                .HasColumnName("name");
+        });
+
+        modelBuilder.Entity<Book>(entity =>
+        {
+            entity.ToTable("book");
+
+            entity.HasIndex(e => e.AuthorId, "IXFK_book_author");
```
Comment (MaltsevaSoftline) [inline LibraryWebAPI/Models/LibraryContext.cs:49]: Do you need this property?

File: LibraryWebAPI/Models/LibraryContext.cs [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+                .HasForeignKey(d => d.OwnerId)
+                .HasConstraintName("FK_book_user");
+        });
+
+        modelBuilder.Entity<Library>(entity =>
+        {
+            entity.ToTable("library");
+
+            entity.Property(e => e.LibraryId)
+                .HasDefaultValue(new Guid())
```
Comment (MaltsevaSoftline) [inline LibraryWebAPI/Models/LibraryContext.cs:93]: Why do you use guid not just numbers as an id?

File: LibraryWebAPI/Models/User.cs [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+
+public partial class User
+{
+    public Guid UserId { get; set; }
+
+    public string UserName { get; set; } = null!;
+
+    public string Password { get; set; } = null!;
+
+    public bool? IsAdmin { get; set; }
```
Comment (MaltsevaSoftline) [inline LibraryWebAPI/Models/User.cs:14]: What if we add another role? If many?

File: LibraryWebAPI/Program.cs [hunk 12]
```diff
@@ -1,12 +1,41 @@
...
+using LibraryWebAPI.Services.BookService;
+using LibraryWebAPI.Services.LibraryService;
+using LibraryWebAPI.Services.UserService;
+
 var builder = WebApplication.CreateBuilder(args);
 
 // Add services to the container.
 
 builder.Services.AddControllers();
+builder.Services.Configure<AuthOptions>(builder.Configuration.GetSection("Auth"));
```
Comment (MaltsevaSoftline) [inline LibraryWebAPI/Program.cs:14]: Cool, like this pattern

File: LibraryWebAPI/Controllers/LibraryController.cs [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+        {
+            var result = await _libraryService.GetLibraryById(libraryId);
+            if (result is null)
+                return NotFound("Hero not found.");
+
+            return Ok(result);
+        }
+
+        // PUT: api/Library/guid
+        [HttpPut("{libraryId}")]
```
Comment (MaltsevaSoftline) [inline LibraryWebAPI/Controllers/LibraryController.cs:43]: We may call this endpoint without admin role and without being authorized at all.

File: LibraryWebAPI/Services/AuthService/AuthService.cs [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+            var authParams = _authOptions.Value;
+
+            var securityKey = authParams.GetSymmetricSecurityKey();
+            var credentials = new SigningCredentials(securityKey, SecurityAlgorithms.HmacSha256);
+
+            var claims = new List<Claim>()
+            {
+                new Claim(JwtRegisteredClaimNames.Name, user.UserName),
+                new Claim(JwtRegisteredClaimNames.Sub, user.UserId.ToString()),
+                new Claim("isAdmin", user.IsAdmin == true ? "yes": "no")
```
Comment (MaltsevaSoftline) [inline LibraryWebAPI/Services/AuthService/AuthService.cs:40]: What is we add a new role? If many?

File: LibraryWebAPI/Services/BookService/BookService.cs [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+            {
+                BookId = Guid.NewGuid(),
+                Name = book.Name,
+                AuthorId = author.Result.AuthorId,
+                Description = book.Description,
+                IsBooked = book.IsBooked,
+                OwnerId = book.OwnerId,
+                BookedDate = book.BookedDate,
+                LibraryId = book.LibraryId,
+            };
```
Comment (MaltsevaSoftline) [inline LibraryWebAPI/Services/BookService/BookService.cs:33]: Add mapper

File: LibraryWebAPI/Services/BookService/BookService.cs [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+        public async Task<List<BookDTO>> DeleteBook(Guid bookId)
+        {
+            var deletedBook = await _context.Books.FindAsync(bookId);
+            if (deletedBook is null)
+                return null;
+
+            var libraryId = deletedBook.LibraryId;
+
+            deletedBook.IsDeleted = true;
+            _context.Entry(deletedBook).State = EntityState.Modified;
```
Comment (MaltsevaSoftline) [inline LibraryWebAPI/Services/BookService/BookService.cs:51]: Why do you need this line?

File: LibraryWebAPI/Services/BookService/BookService.cs [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+                    BookedDate = book.BookedDate,
+                    LibraryId = book.LibraryId,
+                });
+            }
+            return result;
+        }
+
+        public Task<BookDTO?> GetBookById(Guid bookId)
+        {
+            throw new NotImplementedException();
```
Comment (MaltsevaSoftline) [inline LibraryWebAPI/Services/BookService/BookService.cs:99]: ?

File: LibraryWebAPI/Services/LibraryService/LibraryService.cs [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+
+        public async Task<Library?> AddLibrary(Library library)
+        {
+            _context.Libraries.Add(library);
+            await _context.SaveChangesAsync();
+            return await _context.Libraries.FindAsync(library.LibraryId);
+        }
+
+        public async Task<List<Library>> DeleteLibrary(Guid libraryId)
+        {
```
Comment (MaltsevaSoftline) [inline LibraryWebAPI/Services/LibraryService/LibraryService.cs:22]: Is it correct to return entity mode (not dto) to representation level?

File: LibraryWebAPI/Services/UserService/UserService.cs [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+
+        public User? GetUserByLoginData(Login login) =>
+            _context.Users.SingleOrDefault(u =>
+                (u.IsDeleted == false)
+                && (u.UserName == login.UserName) 
+                && (u.Password == ComputeSHA256(login.Password)));
+    
+           
+
+        private string ComputeSHA256(string input)
```
Comment (MaltsevaSoftline) [inline LibraryWebAPI/Services/UserService/UserService.cs:51]: As for me it is auth service logics not user service

File: LibraryWebAPI/Controllers/BookController.cs [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+
+            return NoContent();
+        }
+
+        // POST: api/Book
+        [HttpPost]
+        public async Task<ActionResult<BookDTO>> AddBook(BookDTO book)
+        {
+            await _bookService.AddBook(book);
+            if (book != null)
```
Comment (MaltsevaSoftline) [inline LibraryWebAPI/Controllers/BookController.cs:64]: If book (method parameter) not null?

File: LibraryWebAPI/Controllers/LibraryController.cs [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+        public async Task<ActionResult<Library>> UpdateLibrary(Guid libraryId, Library request)
+        {
+            var result = await _libraryService.UpdateLibrary(libraryId, request);
+            if (result is null)
+                return NotFound("Hero not found.");
+
+            return Ok(result);
```
Comment (MaltsevaSoftline) [inline LibraryWebAPI/Controllers/LibraryController.cs:50]: You return NoContent when update book entity. It's not consistent to implement different behavior in practically similar situations

File: LibraryWebAPI/Services/BookService/BookService.cs [hunk 22]
```diff
@@ -0,0 +1,123 @@
...
+
+        public BookService(LibraryContext context, IAuthorService authorService)
+        {
+            this._context = context;
+            _authorService = authorService;
+        }
+
+        public async Task<BookDTO?> AddBook(BookDTO book)
+        {
+            var author = _authorService.AddAuthor(new Author { AuthorId = Guid.NewGuid(), Name = book.AuthorName });
```
Comment (MaltsevaSoftline) [inline LibraryWebAPI/Services/BookService/BookService.cs:21]: Each author has uniq id?

File: LibraryWebAPI/Services/BookService/BookService.cs [hunk 23]
```diff
@@ -0,0 +1,123 @@
...
+            }
+            return result;
+        }
+
+        public Task<BookDTO?> GetBookById(Guid bookId)
+        {
+            throw new NotImplementedException();
+        }
+
+        public async Task<BookDTO?> UpdateBook(Guid bookId, BookDTO request)
```
Comment (MaltsevaSoftline) [inline LibraryWebAPI/Services/BookService/BookService.cs:102]: BookDto request?

**Review by MaltsevaSoftline — CHANGES_REQUESTED** [review]

File: LibraryWebAPI/GlobalUsings.cs [hunk 24]
```diff
@@ -1,5 +1,17 @@
...
+//Microsoft usings
+global using Microsoft.EntityFrameworkCore;
+global using Microsoft.AspNetCore.Mvc;
+
+//WebLibraryAPI usings
 global using LibraryWebAPI.Common;
 global using LibraryWebAPI.Models;
 global using LibraryWebAPI.Models.DTO;
+global using LibraryWebAPI.Models.DB;
+global using LibraryWebAPI.Helpers;
```
Comment (TatsianaMaltseva) [inline LibraryWebAPI/GlobalUsings.cs:17]: I do not think all you application part should be aware of DB models (especially), or any other models, there are no namespaces on lines 13-17 that must be included if each other file

File: LibraryWebAPI/Helpers/CryptographyHelper.cs [hunk 25]
```diff
@@ -0,0 +1,69 @@
...
+using System.Security.Claims;
+using System.Security.Cryptography;
+using Microsoft.Extensions.Options;
+using Microsoft.IdentityModel.Tokens;
+
+namespace LibraryWebAPI.Helpers
+{
+    public class CryptographyHelper : ICryptographyHelper
+    {
+        private const int MAXIMUM_SALT_LENGTH = 8;
```
Comment (TatsianaMaltseva) [inline LibraryWebAPI/Helpers/CryptographyHelper.cs:11]: Maximum? Your salt length may change?

File: LibraryWebAPI/Models/DB/Author.cs [hunk 26]
```diff
@@ -0,0 +1,10 @@
+﻿namespace LibraryWebAPI.Models.DB;
+
+public partial class Author
+{
+    public Guid AuthorId { get; set; }
```
Comment (TatsianaMaltseva) [inline LibraryWebAPI/Models/DB/Author.cs:5]: AuthorId inside Author entity

File: LibraryWebAPI/Models/DB/AuthorBook.cs [hunk 27]
```diff
@@ -0,0 +1,12 @@
+﻿namespace LibraryWebAPI.Models.DB;
+
+public partial class AuthorBook
+{
+    public Guid AuthorBookId { get; set; }
+    public Guid AuthorId { get; set; }
+    public Guid BookId { get; set; }
+    public bool IsDeleted { get; set; }
+
+    public virtual Author Author { get; set; } = null!;
```
Comment (TatsianaMaltseva) [inline LibraryWebAPI/Models/DB/AuthorBook.cs:10]: Do you use Author and Book property?

File: LibraryWebAPI/Models/DB/Book.cs [hunk 28]
```diff
@@ -0,0 +1,18 @@
+﻿namespace LibraryWebAPI.Models.DB;
+
+public partial class Book
+{
+    public Guid BookId { get; set; }
+    public string Name { get; set; } = null!;
+    public string? Description { get; set; }
+    public byte[]? Cover { get; set; }
```
Comment (TatsianaMaltseva) [inline LibraryWebAPI/Models/DB/Book.cs:8]: Btw do you plan to store binaries in db?

File: LibraryWebAPI/Models/DB/WebLibraryDbContext.cs [hunk 29]
```diff
@@ -0,0 +1,100 @@
...
+
+        modelBuilder.Entity<AuthorBook>(entity =>
+        {
+            entity.ToTable("Author_Book");
+            entity.HasQueryFilter(ab => EF.Property<bool>(ab, "IsDeleted") == false);
+
+            entity.HasOne(d => d.Author).WithMany(p => p.AuthorBooks)
+                .HasForeignKey(d => d.AuthorId)
+                .OnDelete(DeleteBehavior.ClientSetNull)
+                .HasConstraintName("FK_Author_Book_Author");
```
Comment (TatsianaMaltseva) [inline LibraryWebAPI/Models/DB/WebLibraryDbContext.cs:34]: Do you need this line?

File: LibraryWebAPI/Models/DB/WebLibraryDbContext.cs [hunk 30]
```diff
@@ -0,0 +1,100 @@
...
+
+            entity.HasOne(d => d.Book).WithMany(p => p.AuthorBooks)
+                .HasForeignKey(d => d.BookId)
+                .OnDelete(DeleteBehavior.ClientSetNull)
+                .HasConstraintName("FK_Author_Book_Book");
```
Comment (TatsianaMaltseva) [inline LibraryWebAPI/Models/DB/WebLibraryDbContext.cs:39]: and this

File: LibraryWebAPI/Models/DB/WebLibraryDbContext.cs [hunk 31]
```diff
@@ -0,0 +1,100 @@
...
+
+            entity.HasOne(d => d.Owner).WithMany(p => p.Books)
+                .HasForeignKey(d => d.OwnerId)
+                .HasConstraintName("FK_Book_User");
+        });
+
+        modelBuilder.Entity<Library>(entity =>
+        {
+            entity.ToTable("Library");
+            entity.HasQueryFilter(library => EF.Property<bool>(library, "IsDeleted") == false);
```
Comment (TatsianaMaltseva) [inline LibraryWebAPI/Models/DB/WebLibraryDbContext.cs:64]: What if you want to return deleted books also? Is this QueryFilter universal for possible cases regarding this project?

File: LibraryWebAPI/Models/DTO/UserDTO.cs [hunk 32]
```diff
@@ -3,8 +3,8 @@
     public class UserDTO
     {
         public Guid UserId { get; set; }
-        public string UserName { get; set; } = "";
-        public string Password { get; set; } = "";
-        public bool IsAdmin { get; set; } = false;
+        public string UserName { get; set; } = null!;
+        public string Password { get; set; } = null!;
+        public byte UserRole { get; set; }
```
Comment (TatsianaMaltseva) [inline LibraryWebAPI/Models/DTO/UserDTO.cs:8]: Byte? Not enum?

File: LibraryWebAPI/Services/AuthorService/AuthorService.cs [hunk 33]
```diff
@@ -1,17 +1,43 @@
...
+        public async Task<bool> DeleteAuthorAsync(Guid id)
+        {
+            var author = await _context.Authors.FindAsync(id);
+            if (author is null)
+                return false;
+
+            author.IsDeleted = true;
+
+            //Besides deleted author, i should delete AuthorBooks rows, which connected with deleted author
+            foreach (var a in _context.AuthorBooks.Where(ab => ab.AuthorId == author.AuthorId).ToList())
```
Comment (TatsianaMaltseva) [inline LibraryWebAPI/Services/AuthorService/AuthorService.cs:34]: Is it possible to do without foreach? Like for ef do delete it on one's own?

File: LibraryWebAPI/Services/BookService/BookService.cs [hunk 34]
```diff
@@ -1,123 +1,97 @@
...
+            var updBook = await _context.Books.FindAsync(id);
+            if (updBook is null)
+                throw new Exception("Library not found");
 
-            var libraryId = deletedBook.LibraryId;
+            var updBookMapped = _mapper.Map<Book>(book);
 
-            deletedBook.IsDeleted = true;
-            _context.Entry(deletedBook).State = EntityState.Modified;
+            //Nedded cause i can't updBook by "= _mapper.Map()"
```
Comment (TatsianaMaltseva) [inline LibraryWebAPI/Services/BookService/BookService.cs:50]: ??

File: LibraryWebAPI/Services/UserService/UserService.cs [hunk 35]
```diff
@@ -1,65 +1,71 @@
...
             this._context = context;
+            this._cryptoHelper = cryptoHelper;
+            this._mapper = mapper;
         }
 
-        public async Task<List<User>> AddUser(UserDTO userDTO)
+        public async Task<UserDTO> AddUserAsync(UserDTO user)
         {
-            var newUser = new User
+            if (IsUserExist(user.UserName))
```
Comment (TatsianaMaltseva) [inline LibraryWebAPI/Services/UserService/UserService.cs:19]: UserName is uniq?

File: LibraryWebAPI/Services/UserService/UserService.cs [hunk 36]
```diff
@@ -1,65 +1,71 @@
...
-        public User? GetUserByLoginData(Login login) =>
-            _context.Users.SingleOrDefault(u =>
-                (u.IsDeleted == false)
-                && (u.UserName == login.UserName) 
-                && (u.Password == ComputeSHA256(login.Password)));
-    
-           
-
-        private string ComputeSHA256(string input)
+        public User? GetUserByLoginDTO(LoginDTO login)
```
Comment (TatsianaMaltseva) [inline LibraryWebAPI/Services/UserService/UserService.cs:45]: In this metod in 2 of 3 cases you return null

File: LibraryWebAPI/Services/BookService/BookService.cs [hunk 37]
```diff
@@ -1,123 +1,97 @@
...
-            book.AuthorId = _context.Authors.First(au => au.Name == request.AuthorName).AuthorId;
-            book.Description = request.Description;
-            book.IsBooked = request.IsBooked;
-            book.OwnerId = request.OwnerId;
-            book.BookedDate = request.BookedDate;
-            book.LibraryId = request.LibraryId;
-
-            _context.Entry(book).State = EntityState.Modified;
+                //Besides deleted book, i should delete AuthorBooks rows, which connected with deleted book
+                foreach (var a in _context.AuthorBooks.Where(ab => ab.BookId == book.BookId).ToList())
```
Comment (TatsianaMaltseva) [inline LibraryWebAPI/Services/BookService/BookService.cs:88]: Use linq not foreach. Also, do you need to materializе the collection to change i?

**Review by TatsianaMaltseva — CHANGES_REQUESTED** [review]

File: LibraryWebAPI/Models/DB/WebLibraryDbContext.cs (hunk 31 above)
Comment (UnfforU) [reply LibraryWebAPI/Models/DB/WebLibraryDbContext.cs:64]: In my own, I don't think, that it's good idea to get deleted objects by main controllers. 
Soft delete just get u opportunity to fast revert some accidental changes with t-sql, not with API server.

File: LibraryWebAPI/Services/UserService/UserService.cs (hunk 35 above)
Comment (UnfforU) [reply LibraryWebAPI/Services/UserService/UserService.cs:19]: yes

Comment (MaltsevaSoftline) [reply LibraryWebAPI/Services/UserService/UserService.cs:24]: Ok

**Review by MaltsevaSoftline — APPROVED** [review]

**Pull request merged**